samples/custom_dispatch/cpu/embedded/README.md

```diff
@@ -0,0 +1,145 @@
+# Custom CPU Dispatch Functions for Statically-linked Embedded ELFs
+
+See the [custom_dispatch README](/samples/custom_dispatch/README.md) for an
+overview of this approach.
+
+This sample demonstrates how to define external device functions that can be
+dispatched from within IREE programs via simple function calls. Here the
+functions are declared in the MLIR executables, called as normal calls, and
+then defined in a .c file that is cross-compiled for various architectures.
+The compiler uses the attribute specifying which object files to link against
+when performing its final linking and runs LTO to optimize across both the
+generated portions and hand-authored portions.
+
+### Work in Progress
+
+The calling convention used for passing pointers is currently a mess as the MLIR
+`memref` type is used to model buffer references and that expands to many
+arguments. Future revisions will just pass the pointers instead.
+
+Currently weak linkage is not available and external functions must always be
+provided when referenced. In future versions fallback IR will allow for object
+files to be specified only for certain platforms while allowing others to be
+generated via normal codegen paths.
+
+## Workflow
+
+```
++-------------+               +---------------------+       +--------------+
+| functions.c | -> clang -+-> | functions_aarch64.o | -+    | example.mlir |
++-------------+           |   +---------------------+  |    +--------------+
+                          |   +---------------------+  |           v
+                          +-> | functions_x86_64.o  | -+----> iree-compile
+                              +---------------------+              v
+                                                            +--------------+
+                                                            | example.vmfb |
+                                                            +--------------+
+```
+
+1. The user authors their functions in bare-metal C (no TLS, no threads, no
+   malloc, etc). These functions can cover entire workgroups (and a dispatch can
+   be a single workgroup so effectively just function calls) or be utilities
+   used by the function for localized work (microkernels, data type conversion,
+   etc). It's important to remember that parallelism scheduling is done
+   _outside_ of the function via the workgroup count and multiple threads may be
+   executing the function at any time.
```

Comment:
This has me wondering about how those functions can be tested / iterated on. Is that something we could provide a tool for? (in full generality we're leaving that up to users, but it can help to give a bit more of a suggested framework for developers to work within)